datafusion/physical-optimizer/src/limit_pushdown.rs

```diff
@@ -190,7 +194,8 @@ pub fn pushdown_limit_helper(
 
     let skip_and_fetch = Some(global_fetch + global_state.skip);
 
-    if pushdown_plan.supports_limit_pushdown() {
+    let global_skip = global_state.skip;
```

Comment:
Thank you very much for revealing this!

In this `LimitPushdown` optimizer we remove all the `LimitExec`s assuming they might be redundant, assigning them to global_state and then optimizing the plan, and re-adding them if necessary. But in this case, it looks like we forgot to re-add when we're adding the fetch info to SPM.

So rather than overriding `supports_limit_pushdown` API, I think we should handle this where it's happenning. What would you think, solving this with below suggestion?

If you can provide me write permission to your repository I can also commit this change.

```suggestion
    if pushdown_plan.supports_limit_pushdown() {
        if !combines_input_partitions(&pushdown_plan) {
            // We have information in the global state and the plan pushes down,
            // continue:
            Ok((Transformed::no(pushdown_plan), global_state))
        } else if let Some(plan_with_fetch) = pushdown_plan.with_fetch(skip_and_fetch) {
            // This plan is combining input partitions, so we need to add the
            // fetch info to plan if possible. If not, we must add a `LimitExec`
            // with the information from the global state.
            let mut new_plan = plan_with_fetch;
            if global_state.skip > 0 {
                new_plan =
                    add_global_limit(new_plan, global_state.skip, global_state.fetch);
            }
            global_state.fetch = skip_and_fetch;
            global_state.skip = 0;
            global_state.satisfied = true;
            Ok((Transformed::yes(new_plan), global_state))
        } else if global_state.satisfied {
```